app/src/main/kotlin/net/rafaeltoledo/social/data/auth/GoogleAuth.kt

```diff
@@ -0,0 +1,51 @@
+package net.rafaeltoledo.social.data.auth
+
+import android.app.Activity
+import android.content.Intent
+import com.google.android.gms.auth.api.signin.GoogleSignIn
+import com.google.android.gms.auth.api.signin.GoogleSignInClient
+import com.google.android.gms.auth.api.signin.GoogleSignInOptions
+import com.google.android.gms.common.api.ApiException
+import net.rafaeltoledo.social.BuildConfig
+
+class GoogleAuth : DelegatedAuth {
+
+    private var client: GoogleSignInClient? = null
+
+    override fun <T : DelegatedAuth> build(activity: Activity): T {
+        if (client != null) {
+            @Suppress("UNCHECKED_CAST") return this as T
+        }
+
+        val options = GoogleSignInOptions.Builder()
+                .requestIdToken(BuildConfig.GOOGLE_REQUEST_ID_TOKEN)
+                .requestEmail()
+                .build()
+
+        client = GoogleSignIn.getClient(activity, options)
+
+        @Suppress("UNCHECKED_CAST") return this as T
+    }
+
+    override fun signIn(activity: Activity) {
+        activity.startActivityForResult(client?.signInIntent, RESULT_SIGN_IN)
+    }
+
+    override fun onResult(requestCode: Int, resultCode: Int, data: Intent?): AuthResult {
+        val task = GoogleSignIn.getSignedInAccountFromIntent(data)
+        return try {
+            val account = task.getResult(ApiException::class.java)
+            AuthResult(Status.SUCCESS, account.idToken)
+        } catch (e: ApiException) {
+            AuthResult(Status.FAILURE)
+        }
+    }
+
+    override fun signOut(callback: (Status) -> Unit) {
+        client?.signOut()?.addOnCompleteListener { callback.invoke(if (it.isSuccessful) Status.SUCCESS else Status.FAILURE) }
+    }
+
+    companion object {
+        const val RESULT_SIGN_IN = 1441
```

Comment:
Just some dumbness from my part here... What is the convention here nowadays? Can't you use `const val` outside the class declaration on this file? It would be on a companion object with the package name, right?

Reply:
There's no hard convention here, but I usually see constants inside a companion object ([example](https://github.com/android/android-ktx/blob/3b47c7561c498c4e6be91fa46987f88f7e795a8e/src/androidTest/java/androidx/core/TestPreferenceActivity.kt)).